src/Interpreters/QueryMetricLog.cpp

```diff
@@ -0,0 +1,201 @@
+        auto current_time = std::chrono::system_clock::now();
+        const auto query_info = process_list.getQueryInfo(query_id, false, true, false);
+        if (!query_info)
+        {
+            LOG_TRACE(logger, "Query {} is not running anymore, so we couldn't get its QueryInfo", query_id);
+            return;
+        }
+
+        auto elem = createLogMetricElement(query_id, *query_info, current_time);
+        if (elem)
+            add(std::move(elem.value()));
+    });
+
+    status.task->scheduleAfter(interval_milliseconds);
+
+    std::lock_guard lock(queries_mutex);
+    queries.emplace(query_id, std::move(status));
+}
+
+void QueryMetricLog::finishQuery(const String & query_id, QueryStatusInfoPtr query_info)
+{
+    std::unique_lock lock(queries_mutex);
+    auto it = queries.find(query_id);
+
+    /// finishQuery may be called from logExceptionBeforeStart when the query has not even started
+    /// yet, so its corresponding startQuery is never called.
+    if (it == queries.end())
+        return;
+
+    if (query_info)
+    {
+        auto elem = createLogMetricElement(query_id, *query_info, std::chrono::system_clock::now(), false);
+        if (elem)
+            add(std::move(elem.value()));
+    }
+
+    /// The task has an `exec_mutex` locked while being executed. This same mutex is locked when
+    /// deactivating the task, which happens automatically on its destructor. Thus, we cannot
+    /// deactivate/destroy the task while it's running. Now, the task locks `queries_mutex` to
+    /// prevent concurrent edition of the queries. In short, the mutex order is: exec_mutex ->
+    /// queries_mutex. Thus, to prevent a deadblock we need to make sure that we always lock them in
+    /// that order.
+    {
+        /// Take ownership of the task so that we can destroy it in this scope after unlocking `queries_lock`.
+        auto task = std::move(it->second.task);
+
+        /// Build an empty task for the old task to make sure it does not lock any mutex on its destruction.
+        it->second.task = {};
+
+        /// Ensure `queries_mutex` is unlocked before calling task's destructor at the end of this
+        /// scope which will lock `exec_mutex`.
+        lock.unlock();
+    }
+
+    lock.lock();
+    queries.erase(query_id);
+}
+
+std::optional<QueryMetricLogElement> QueryMetricLog::createLogMetricElement(const String & query_id, const QueryStatusInfo & query_info, TimePoint current_time, bool schedule_next)
+{
+    LOG_DEBUG(logger, "Collecting query_metric_log for query {}. Schedule next: {}", query_id, schedule_next);
+    std::lock_guard lock(queries_mutex);
+    auto query_status_it = queries.find(query_id);
+
+    /// The query might have finished while the scheduled task is running.
+    if (query_status_it == queries.end())
+    {
+        LOG_TRACE(logger, "Query {} finished already while this collecting task was running", query_id);
```

Comment:
This log is still under the mutex :cry:

Reply:
Done! [`2a1327d` (#66532)](https://github.com/ClickHouse/ClickHouse/pull/66532/commits/2a1327dd91a05f8367e6bf51ae76a03c0ae9f4c7)